op-proposer/proposer/driver.go

```diff
@@ -77,15 +77,19 @@ type L2OutputSubmitter struct {
 }
 
 // NewL2OutputSubmitter creates a new L2 Output Submitter
-func NewL2OutputSubmitter(setup DriverSetup) (*L2OutputSubmitter, error) {
+func NewL2OutputSubmitter(setup DriverSetup) (_ *L2OutputSubmitter, err error) {
 	ctx, cancel := context.WithCancel(context.Background())
+	defer func() {
+		if err != nil || recover() != nil {
+			cancel()
+		}
+	}()
 
 	if setup.Cfg.L2OutputOracleAddr != nil {
 		return newL2OOSubmitter(ctx, cancel, setup)
 	} else if setup.Cfg.DisputeGameFactoryAddr != nil {
 		return newDGFSubmitter(ctx, cancel, setup)
```

Comment:
Why not just have the `newL2OOSubmitter` and `newDGFSubmitter` methods create their own contexts respectively?

Reply:
Unless we need to pass the runtime context in tests or so to the constructor, this might actually be cleaner. Then the constructor owns the creation or the fields it creates.

Reply:
We might then still end up either doing the
```
defer func() {
		if err != nil || recover() != nil {
			cancel()
		}
	}()
```
construction in the ctors, or just call `cancel()` separately on each error, ignoring panics.

Reply:
I see it's ticketed at https://github.com/ethereum-optimism/client-pod/issues/443